### PR TITLE
add fs tab for kbfs team

### PR DESCRIPTION
**Pull request by chrisnojima**
@keybase/react-hackers this adds a new tab (feature flagged to admins) for the fs components (new kbfs in-app finder)

cc: @jzila

### CODE REVIEW

**Comment by jzila** [comment]
This is very helpful, thanks!
cc @taruti, @strib, @songgao 

side note: we need an @keybase/kbfs-hackers

File: shared/fs/index.stories.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// @flow
```
Comment (jzila) [inline shared/fs/index.stories.js]: Is this the file we'd play with for StoryBook?

Also, I noticed that you were running that as a server on your localhost. How do we access it?

**Comment by chrisnojima** [comment]
Yup. Just run 'yarn storybook' and it launches the local server

File: shared/reducers/fs.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+import * as Types from '../constants/types/fs'
+
+const initialState = Constants.makeState()
+
+export default function(state: Types.State = initialState, action: FSGen.Actions) {
+  switch (action.type) {
+    case FSGen.resetStore:
+      return initialState
+    case FSGen.increaseCount: {
+      const {amount = 1} = action.payload
```
Comment (jzila) [inline shared/reducers/fs.js]: What's going on here? Having trouble detangling this syntax.

File: shared/reducers/fs.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+      return state.withMutations(s => s.set('counter', state.counter + amount))
+    }
+    default:
+      // eslint-disable-next-line no-unused-expressions
+      ;(action: empty) // if you get a flow error here it means there's an action you claim to handle but didn't
```
Comment (jzila) [inline shared/reducers/fs.js]: Same--what does this syntax do?

File: shared/reducers/fs.js (hunk 2 above)
Comment (chrisnojima) [reply shared/reducers/fs.js]: sure. this is destructuring (https://babeljs.io/learn-es2015/#ecmascript-2015-features-destructuring)
meaning this is basically
```
const amount = action.payload === undefined ? 1 : action.payload
```

You can test all these new features here: https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=MYewdgzgLgBA3gQwLYgK5lgXhgRgL4zYLBQCW4AdAA4ICeANiAgCZA&debug=false&circleciRepo=&evaluate=true&fileSize=false&lineWrap=false&presets=es2015%2Creact%2Cstage-0%2Cstage-1%2Cstage-2%2Cstage-3%2Ces2015-loose%2Cenv&prettier=false&targets=&version=6.26.0

File: shared/reducers/fs.js (hunk 3 above)
Comment (chrisnojima) [reply shared/reducers/fs.js]: this is basically to get flow to help us ensure we handle all cases in the switch statement.

ala: https://github.com/facebook/flow/issues/2695#issuecomment-256845109

this is a force cast of the action to a type that doesn't exist so it'll complain. try and comment out a case and run flow and it'll complain
`yarn flow`

Comment (jzila) [reply shared/reducers/fs.js]: In the examples there's no `;`, is that needed?

Comment (chrisnojima) [reply shared/reducers/fs.js]: prettier is injecting that in automatically

Comment (jzila) [reply shared/reducers/fs.js]: I see, thanks.

File: shared/constants/fs.js [hunk 4]
```diff
@@ -0,0 +1,7 @@
+// @flow
+import * as I from 'immutable'
+import * as Types from './types/fs'
+
+export const makeState: I.RecordFactory<Types._State> = I.Record({
+  counter: 0,
```
Comment (songgao) [inline shared/constants/fs.js]: Is this where we'll populate default states for the redux store?

File: shared/fs/container.js [hunk 5]
```diff
@@ -0,0 +1,16 @@
+// @flow
+import {connect, type TypedState, type Dispatch} from '../util/container'
+import * as FSGen from '../actions/fs-gen'
+import Fs from '.'
+
+const mapStateToProps = (state: TypedState) => ({
```
Comment (songgao) [inline shared/fs/container.js]: Just to make sure I understand correctly, this is called whenever *any* state is updated in from Redux. Then if the object (prop) returned from this function is identical to the previous prop, it won't trigger a render. But if e.g. counter changes, it triggers a render. Does this sound correct?

File: shared/fs/container.js [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  counter: state.fs.counter,
+  you: state.config.username,
+})
+
+const mapDispatchToProps = (dispatch: Dispatch) => ({
+  increase: () => dispatch(FSGen.createIncreaseCount({amount: 1})),
```
Comment (songgao) [inline shared/fs/container.js]: Does `dispatch(...)` here return an object to be merged later? How is `dispatch` implemented? Also, is `increase` and `increase10` like event names? I see they are functions on the component props. Does redux map them automatically?

File: shared/reducers/fs.js (hunk 3 above)
Comment (songgao) [reply shared/reducers/fs.js]: Cool hack! Why is the semicolon needed here? I see linked GH comment doesn't have it.

Comment (songgao) [reply shared/reducers/fs.js]: Oops nvm just say jzila asked the same question above.

**Comment by chrisnojima** [comment]
I highly recommend looking at the code under devices so you can see a fully fleshed out structure. Questions here are fine but this is very bare bones code

File: shared/constants/fs.js (hunk 4 above)
Comment (chrisnojima) [reply shared/constants/fs.js]: Sort of. Really this happens in the reducer. It just so happens to take this value. The reducer is called with undefined as the initial state

File: shared/fs/container.js (hunk 5 above)
Comment (chrisnojima) [reply shared/fs/container.js]: Also sort of. First it's a shallow equal compare so the entire object doesn't need to be the same. If all the properties are the same then connect will avoid the render. There are things that can subvert this but this is the common case

File: shared/fs/container.js (hunk 6 above)
Comment (chrisnojima) [reply shared/fs/container.js]: Dispatch comes from redux itself. How it's handled is dependent on how you configure redux. You should watch the video I mention in the ticket

**Pull request merged**